### ISSUE TITLE
Raise windows while cycling through them in Alt-Tab

### ISSUE THREAD

**Issue by danfe** [bbidulock/icewm#564]
I'm looking for QuickSwitch configuration option which would raise windows while cycling through them with Alt-Tab, and don't seem to find one.  Currently, to-be-focused (candidate) window is only indicated by the frame (active vs. inactive), but the window is not raised until Alt-Tab is released.  This is usually not a problem when there are few windows on the screen and they don't overlap (cover) others completely, but try launching a dozen of xterms and that's when raise-while-cycling becomes really useful.

I've patched my copy of IceWM like this, but that's probably wrong, it's just a PoC so you can understand what I mean better:
```diff
--- src/wmswitch.cc.orig
+++ src/wmswitch.cc
@@ -119,6 +119,7 @@ class WindowItemsCtrlr : public ISwitchItems
 
     void displayFocusChange(YFrameWindow *frame)  {
         manager->switchFocusTo(frame, false);
+        frame->wmRaise();
     }
 
 public:

```

**Comment by gijsbers** [comment]
Yeah, this function is the reason for #558, so your addition won't help.
I think it shouldn't change the clients WM stacking order,
or very carefully restore it when the next tab or a cancel occurs.
Maybe easier would be to only temporarily call XRestackWindows
to restack the window just below the QuickSwitch overlay.

**Comment by danfe** [comment]
> I think it shouldn't change the clients WM stacking order, or very carefully restore it when the next tab or a cancel occurs.

I see.  Alternative to window raising would be any kind of indication that would help to identify the particular window during cycling over them.  Openbox does not raise windows, but it draws a rather fat black-and-white frame, perhaps IceWM could do something similar, but less ugly and themable (AFAICT, Openbox does not allow to adjust thickness or colors of it, which kind of sucks).

**Comment by qsmodo** [comment]
Did you know you can set the Xterm's title to show your last issued command with a `DEBUG` trap? [Display command in xterm titlebar](https://unix.stackexchange.com/q/9007). That does not solve the issue in the general case, but still you may find it helpful.

**Comment by gijsbers** [comment]
Could you verify that commit 869d42c suffices?

**Comment by danfe** [comment]
> Could you verify that commit 869d42c suffices?

Hmm, I guess it kind of works, but since you've mentioned the window stacking order, I've now started to pay closer attention, and noticed that if I switch between two workspaces back and forth while in Alt-Tab, window order would not always persist: it seems that two topmost windows would be constantly changing their relative position every time.  Not sure if that's intended, but I guess it's okay.

FWIW, Openbox always displays identical window lists (different for each workspace, but they remain unchanged between the switches), but then again, it does not raise windows during selection and only gives focus/raises once selection is made, and does not retain current position in the Alt-Tab windows between workspaces switches.

**Comment by gijsbers** [comment]
Good point! Try again.

**Comment by danfe** [comment]
Yeah, it feels better now, thanks!

**Comment by Code7R** [comment]
This breaks the switching with an arbitrary key in derived mode(s). I.e.

switchkey "Super+p"			icewm-menu-xrandr

Now it gets stuck after first Win-P pressing, only mouse can rotate there. Even cursor keys not working.

**Comment by gijsbers** [comment]
Fixed in 43de75c4085aecacfc2b5613031062c8dcbdbbe3.
